lightning/api_clnraw.go

```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
+		Mili        = 1000
+	)
+
+	var fundsReply ClnFundsChainResp
+	err := l.connection.Call(ctx, ListFunds, []string{}, &fundsReply)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &Funds{}
+	f.ConfirmedBalance = 0
+	f.TotalBalance = 0
+	f.LockedBalance = 0
+
+	for _, one := range fundsReply.Outputs {
+		amount := int64(ConvertAmount(one.AmountMsat))
+		if !one.Reserved && one.Status == Confirmed {
+			f.ConfirmedBalance += (amount / Mili)
+			f.TotalBalance += (amount / Mili)
+		} else if !one.Reserved && one.Status == Unconfirmed {
+			f.TotalBalance += (amount / Mili)
+		} else if one.Reserved && (one.Status == Unconfirmed || one.Status == Confirmed) {
+			f.LockedBalance += (amount / Mili)
+		}
+	}
+
+	return f, nil
+}
+
+// SendToOnChainAddress - API call.
+func (l *ClnRawLightningAPI) SendToOnChainAddress(ctx context.Context, address string, sats int64, useUnconfirmed bool, urgency Urgency) (string, error) {
+	var reply ClnWithdrawResp
+
+	feerate := "normal"
+	switch urgency {
+	case Urgent:
+		feerate = "urgent"
+	case Normal:
+		feerate = "normal"
+	case Low:
+		feerate = "slow" // slow not low, beware!
+	}
+
+	minConf := 1
+	if useUnconfirmed {
+		minConf = 0
+	}
+
+	err := l.connection.Call(ctx, Withdraw, []interface{}{address, sats, feerate, minConf}, &reply)
+	if err != nil {
+		return "", err
+	}
+
+	if len(reply.TxID) == 0 {
+		return "", fmt.Errorf("invalid transaction id")
+	}
+
+	return reply.TxID, nil
+}
+
+// calculateExclusion is used since CLN can just exclude channels, so we tell it which channels we want and method calculates the "inverse" of it
+func (l *ClnRawLightningAPI) calculateExclusion(ctx context.Context, outgoingChanIds []uint64) ([]string, error) {
+	result := make([]string, 0)
+
+	if outgoingChanIds == nil || len(outgoingChanIds) == 0 {
+		return nil, nil
+	}
+
+	var info ClnInfo
+	err := l.connection.Call(ctx, GetInfo, []string{}, &info)
+	if err != nil {
+		return nil, err
+	}
+
+	chans, err := l.getMyChannels(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[uint64]struct{})
+	for _, one := range outgoingChanIds {
+		m[one] = struct{}{}
+	}
+
+	for _, one := range chans {
+		_, ok := m[one.ChannelID]
+		if !ok {
+			// not an outgoing channel, create exclusion
+
+			id := FromLndChanID(one.ChannelID)
+
+			peer := ""
+			if one.Node1Pub == info.PubKey {
+				peer = one.Node2Pub
+			} else if one.Node2Pub == info.PubKey {
+				peer = one.Node1Pub
+			} else {
+				glog.Warningf("Channel %d is not mine?", one.ChannelID)
+				continue
+			}
+
+			// direction (u32): 0 if this channel is traversed from lesser to greater id, otherwise 1
+			direction := 0
+			if info.PubKey > peer {
+				direction = 1
+			}
+
+			result = append(result, fmt.Sprintf("%s/%d", id, direction))
+		}
+	}
+
+	return result, nil
+}
+
+// PayInvoice - API call.
+func (l *ClnRawLightningAPI) PayInvoice(ctx context.Context, paymentRequest string, sats int64, outgoingChanIds []uint64) (*PaymentResp, error) {
```

Comment:
looks like this are `msats` and not `sats` if we need to multiply with 1000 ?

Reply:
Parameter name is correct, we get "sats" but need to multiply by 1000 since CLN API uses msats.

Reply:
Actually this got me thinking, probably an API with milisats is more future proof when 1 sat = 1 cent 😄